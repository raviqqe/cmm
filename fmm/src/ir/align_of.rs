use super::size_of::SizeOf;
use crate::types::{self, Type};

#[derive(Clone, Debug, PartialEq)]
pub struct AlignOf {
    type_: Type,
}

impl AlignOf {
    pub const RESULT_TYPE: types::Primitive = SizeOf::RESULT_TYPE;

    pub fn new(type_: impl Into<Type>) -> Self {
        Self {
            type_: type_.into(),
        }
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }
}
