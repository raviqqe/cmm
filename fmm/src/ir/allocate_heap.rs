use super::expression::Expression;
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq)]
pub struct AllocateHeap(Arc<AllocateHeapInner>);

#[derive(Clone, Debug, PartialEq)]
struct AllocateHeapInner {
    size: Expression,
    name: String,
}

impl AllocateHeap {
    pub fn new(size: impl Into<Expression>, name: impl Into<String>) -> Self {
        Self(
            AllocateHeapInner {
                size: size.into(),
                name: name.into(),
            }
            .into(),
        )
    }

    pub fn size(&self) -> &Expression {
        &self.0.size
    }

    pub fn name(&self) -> &str {
        &self.0.name
    }
}
