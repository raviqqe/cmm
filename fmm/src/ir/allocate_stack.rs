use crate::types::Type;
use std::rc::Rc;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct AllocateStack(Rc<AllocateStackInner>);

#[derive(Clone, Debug, PartialEq, Eq)]
struct AllocateStackInner {
    type_: Type,
    name: String,
}

impl AllocateStack {
    pub fn new(type_: impl Into<Type>, name: impl Into<String>) -> Self {
        Self(
            AllocateStackInner {
                type_: type_.into(),
                name: name.into(),
            }
            .into(),
        )
    }

    pub fn type_(&self) -> &Type {
        &self.0.type_
    }

    pub fn name(&self) -> &str {
        &self.0.name
    }
}
