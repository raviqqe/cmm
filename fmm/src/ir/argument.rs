use super::ArgumentOptions;
use crate::types::Type;

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Argument {
    name: String,
    type_: Type,
    options: ArgumentOptions,
}

impl Argument {
    pub fn new(name: impl Into<String>, type_: impl Into<Type>) -> Self {
        Self::with_options(name, type_, Default::default())
    }

    pub fn with_options(
        name: impl Into<String>,
        type_: impl Into<Type>,
        options: ArgumentOptions,
    ) -> Self {
        Self {
            name: name.into(),
            type_: type_.into(),
            options,
        }
    }

    pub fn name(&self) -> &str {
        &self.name
    }

    pub fn type_(&self) -> &Type {
        &self.type_
    }

    pub fn type_mut(&mut self) -> &mut Type {
        &mut self.type_
    }

    pub fn options(&self) -> &ArgumentOptions {
        &self.options
    }
}
