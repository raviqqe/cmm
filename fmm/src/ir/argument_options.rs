#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ArgumentOptions {
    alias: bool,
}

impl ArgumentOptions {
    pub fn new() -> Self {
        Self { alias: true }
    }

    pub fn alias(&self) -> bool {
        self.alias
    }

    pub fn set_alias(self, alias: bool) -> Self {
        Self { alias, ..self }
    }
}

impl Default for ArgumentOptions {
    fn default() -> Self {
        Self::new()
    }
}
