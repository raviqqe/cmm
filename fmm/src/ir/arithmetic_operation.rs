use super::expression::Expression;
use crate::types;
use std::sync::Arc;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum ArithmeticOperator {
    Add,
    Subtract,
    Multiply,
    Divide,
}

#[derive(Clone, Debug, PartialEq)]
pub struct ArithmeticOperation(Arc<ArithmeticOperationInner>);

#[derive(Debug, PartialEq)]
struct ArithmeticOperationInner {
    type_: types::Primitive,
    operator: ArithmeticOperator,
    lhs: Expression,
    rhs: Expression,
}

impl ArithmeticOperation {
    pub fn new(
        type_: types::Primitive,
        operator: ArithmeticOperator,
        lhs: impl Into<Expression>,
        rhs: impl Into<Expression>,
    ) -> Self {
        Self(
            ArithmeticOperationInner {
                type_,
                operator,
                lhs: lhs.into(),
                rhs: rhs.into(),
            }
            .into(),
        )
    }

    pub fn type_(&self) -> types::Primitive {
        self.0.type_
    }

    pub fn operator(&self) -> ArithmeticOperator {
        self.0.operator
    }

    pub fn lhs(&self) -> &Expression {
        &self.0.lhs
    }

    pub fn rhs(&self) -> &Expression {
        &self.0.rhs
    }
}
