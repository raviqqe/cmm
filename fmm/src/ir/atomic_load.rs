use super::{atomic_ordering::AtomicOrdering, expression::Expression};
use crate::types::Type;

#[derive(Clone, Debug, PartialEq)]
pub struct AtomicLoad(Box<AtomicLoadInner>);

#[derive(Clone, Debug, PartialEq)]
struct AtomicLoadInner {
    type_: Type, // pointer element type
    pointer: Expression,
    ordering: AtomicOrdering,
    name: String,
}

impl AtomicLoad {
    pub fn new(
        type_: impl Into<Type>,
        pointer: impl Into<Expression>,
        ordering: AtomicOrdering,
        name: impl Into<String>,
    ) -> Self {
        Self(
            AtomicLoadInner {
                type_: type_.into(),
                pointer: pointer.into(),
                ordering,
                name: name.into(),
            }
            .into(),
        )
    }

    pub fn type_(&self) -> &Type {
        &self.0.type_
    }

    pub fn pointer(&self) -> &Expression {
        &self.0.pointer
    }

    pub fn ordering(&self) -> AtomicOrdering {
        self.0.ordering
    }

    pub fn name(&self) -> &str {
        &self.0.name
    }
}
