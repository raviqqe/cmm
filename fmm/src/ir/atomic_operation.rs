use super::{atomic_ordering::AtomicOrdering, expression::Expression};
use crate::types;

#[derive(Copy, Clone, Debug, Eq, PartialEq)]
pub enum AtomicOperator {
    Add,
    Subtract,
}

#[derive(Clone, Debug, PartialEq)]
pub struct AtomicOperation(Box<AtomicOperationInner>);

#[derive(Clone, Debug, PartialEq)]
struct AtomicOperationInner {
    type_: types::Primitive,
    operator: AtomicOperator,
    pointer: Expression,
    value: Expression,
    ordering: AtomicOrdering,
    name: String,
}

impl AtomicOperation {
    pub fn new(
        type_: types::Primitive,
        operator: AtomicOperator,
        pointer: impl Into<Expression>,
        value: impl Into<Expression>,
        ordering: AtomicOrdering,
        name: impl Into<String>,
    ) -> Self {
        Self(
            AtomicOperationInner {
                type_,
                operator,
                pointer: pointer.into(),
                value: value.into(),
                ordering,
                name: name.into(),
            }
            .into(),
        )
    }

    pub fn type_(&self) -> types::Primitive {
        self.0.type_
    }

    pub fn type_mut(&mut self) -> &mut types::Primitive {
        &mut self.0.type_
    }

    pub fn operator(&self) -> AtomicOperator {
        self.0.operator
    }

    pub fn pointer(&self) -> &Expression {
        &self.0.pointer
    }

    pub fn pointer_mut(&mut self) -> &mut Expression {
        &mut self.0.pointer
    }

    pub fn value(&self) -> &Expression {
        &self.0.value
    }

    pub fn value_mut(&mut self) -> &mut Expression {
        &mut self.0.value
    }

    pub fn ordering(&self) -> AtomicOrdering {
        self.0.ordering
    }

    pub fn name(&self) -> &str {
        &self.0.name
    }
}
