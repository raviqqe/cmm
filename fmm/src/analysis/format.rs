use crate::ir::*;

pub fn format(module: &Module) -> String {
    format!(
        "(module\n{}\n)",
        module
            .function_definitions()
            .iter()
            .map(|definition| format_function_definition(definition, 1))
            .collect::<Vec<_>>()
            .join("\n")
    )
}

fn format_function_definition(definition: &FunctionDefinition, level: usize) -> String {
    format!(
        "{indent}(function {} {}\n{}\n{indent})",
        definition.name(),
        definition
            .arguments()
            .iter()
            .map(|argument| argument.name())
            .collect::<Vec<_>>()
            .join(" "),
        format_block(definition.body(), level + 1),
        indent = indent(level)
    )
}

fn format_block(block: &Block, level: usize) -> String {
    let instructions = block
        .instructions()
        .iter()
        .map(|instruction| format_instruction(instruction, level + 1))
        .collect::<Vec<_>>()
        .join(" ");

    format!(
        "{indent}(\n{}{}\n{indent})",
        if instructions.is_empty() {
            "".into()
        } else {
            instructions.to_owned() + "\n"
        },
        indent(level + 1) + &format_terminal_instruction(block.terminal_instruction()),
        indent = indent(level)
    )
}

fn format_instruction(instruction: &Instruction, level: usize) -> String {
    indent(level)
        + &match instruction {
            Instruction::Call(call) => format!(
                "(call {} {})",
                format_expression(call.function()),
                call.arguments()
                    .iter()
                    .map(|argument| format_expression(argument))
                    .collect::<Vec<_>>()
                    .join(" ")
            ),
            Instruction::If(if_) => format!(
                "(if {}\n{}\n{})",
                format_expression(if_.condition()),
                format_block(if_.then(), level + 1),
                format_block(if_.else_(), level + 1),
            ),
            _ => todo!(),
        }
}

fn format_terminal_instruction(instruction: &TerminalInstruction) -> String {
    match instruction {
        TerminalInstruction::Branch(branch) => {
            format!("(branch {})", format_expression(branch.expression()))
        }
        TerminalInstruction::Return(return_) => {
            format!("(return {})", format_expression(return_.expression()))
        }
        TerminalInstruction::Unreachable => "(unreachable)".into(),
    }
}

fn format_expression(expression: &Expression) -> String {
    match expression {
        Expression::AlignOf(_) => "(align-of)".into(),
        Expression::ArithmeticOperation(operation) => format!(
            "({} {} {})",
            match operation.operator() {
                ArithmeticOperator::Add => "+",
                ArithmeticOperator::Subtract => "-",
                ArithmeticOperator::Multiply => "*",
                ArithmeticOperator::Divide => "/",
            },
            format_expression(operation.lhs()),
            format_expression(operation.rhs())
        ),
        Expression::BitCast(bit_cast) => {
            format!("(bit-cast {})", format_expression(bit_cast.expression()))
        }
        Expression::ComparisonOperation(operation) => format!(
            "({} {} {})",
            match operation.operator() {
                ComparisonOperator::Equal => "==",
                ComparisonOperator::NotEqual => "!=",
                ComparisonOperator::LessThan => "<",
                ComparisonOperator::LessThanOrEqual => "<=",
                ComparisonOperator::GreaterThan => ">",
                ComparisonOperator::GreaterThanOrEqual => ">=",
            },
            format_expression(operation.lhs()),
            format_expression(operation.rhs())
        ),
        Expression::Primitive(primitive) => format_primitive(primitive),
        Expression::SizeOf(_) => "(size-of)".into(),
        Expression::Undefined(_) => "undefined".into(),
        Expression::Variable(variable) => variable.name().into(),
        _ => todo!(),
    }
}

fn format_primitive(primitive: &Primitive) -> String {
    match primitive {
        Primitive::Boolean(boolean) => format!("{}", boolean),
        Primitive::Float32(number) => format!("{}", number),
        Primitive::Float64(number) => format!("{}", number),
        Primitive::Integer8(number) => format!("{}", number),
        Primitive::Integer32(number) => format!("{}", number),
        Primitive::Integer64(number) => format!("{}", number),
        Primitive::PointerInteger(number) => format!("{}", number),
    }
}

fn indent(level: usize) -> String {
    "  ".repeat(level)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types;

    #[test]
    fn format_module_with_call() {
        insta::assert_snapshot!(
            "{}",
            format(&Module::new(
                vec![],
                vec![],
                vec![],
                vec![FunctionDefinition::new(
                    "foo",
                    vec![],
                    Block::new(
                        vec![Call::new(
                            types::Function::new(
                                vec![],
                                types::Primitive::Boolean,
                                types::CallingConvention::Source
                            ),
                            Variable::new("f"),
                            vec![Variable::new("a1").into(), Variable::new("a2").into()],
                            "result",
                        )
                        .into()],
                        Return::new(types::Primitive::Boolean, Primitive::Boolean(true))
                    ),
                    types::Primitive::Boolean,
                    types::CallingConvention::Source,
                    Linkage::Internal
                )]
            ))
        );
    }

    #[test]
    fn format_module_with_if() {
        insta::assert_snapshot!(format(&Module::new(
            vec![],
            vec![],
            vec![],
            vec![FunctionDefinition::new(
                "f",
                vec![],
                Block::new(
                    vec![If::new(
                        types::Primitive::Float64,
                        Primitive::Boolean(true),
                        Block::new(
                            vec![],
                            Branch::new(types::Primitive::Float64, Primitive::Float64(42.0)),
                        ),
                        Block::new(
                            vec![],
                            Branch::new(types::Primitive::Float64, Primitive::Float64(42.0)),
                        ),
                        "x",
                    )
                    .into()],
                    Return::new(types::Primitive::Float64, Variable::new("x")),
                ),
                types::Primitive::Float64,
                types::CallingConvention::Source,
                Linkage::Internal
            )],
        )));
    }
}
