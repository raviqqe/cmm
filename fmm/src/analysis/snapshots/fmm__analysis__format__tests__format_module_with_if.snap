---
source: fmm/src/analysis/format.rs
expression: "format_module(&Module::new(vec![], vec![], vec![],\n                           vec![FunctionDefinition ::\n                                new(\"f\", vec! [], Block ::\n                                    new(vec!\n                                        [If ::\n                                         new(types :: Primitive :: Float64,\n                                             Primitive :: Boolean(true), Block\n                                             ::\n                                             new(vec! [], Branch ::\n                                                 new(types :: Primitive ::\n                                                     Float64, Primitive ::\n                                                     Float64(42.0)),), Block\n                                             ::\n                                             new(vec! [], Branch ::\n                                                 new(types :: Primitive ::\n                                                     Float64, Primitive ::\n                                                     Float64(42.0)),),\n                                             \"x\",).into()], Return ::\n                                        new(types :: Primitive :: Float64,\n                                            Variable :: new(\"x\")),), types ::\n                                    Primitive :: Float64, types ::\n                                    CallingConvention :: Source, Linkage ::\n                                    Internal)]))"

---
(module
  (function f 
    (block
      (if true
        (block
          (branch 42)
        )
        (block
          (branch 42)
        ))
      (return x)
    )
  )
)
